### ISSUE TITLE
"strict mode" compare-struct-to-null doesn't seem to be working.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#14760]
I've made the mistake in Roslyn several times of comparing a struct to null.  Specifically, I've compared ImmutableArray's to null. This often happens when refactoring existing code to move from a reference-collection to ImmutableArray.  These weren't caught in Roslyn even though @gafter believes they should have.

**Comment by CyrusNajmabadi** [comment]
Tagging @gafter about this.

**Comment by AdamSpeight2008** [comment]
I think if we where going to do this it should be a warning, rather than an error.
As in VB.net is it a valid comparison ` mystruct = Nothing ` as it equivalent to,
 ` mystruct == defaultof( mystructType )`

**Comment by CyrusNajmabadi** [comment]
C# semantics are not the same as VB semantics here.

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi In which it should be a warning.

**Comment by CyrusNajmabadi** [comment]
Code that is always contains a bug should probably be an error :)

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi but in legacy code it is valid. A warning would still permit their code to compile and execute. Unless they have warnings as error enabled.

**Comment by gafter** [comment]
@CyrusNajmabadi Code such as
```cs
if (1 == null) {}
```
is legal C# because the language specification allows it (by virtue of the spec-required `operator ==(int?, int?)`, thereby requiring a conforming implementation to compile that code. But we certainly can give a warning.

**Comment by CyrusNajmabadi** [comment]
I mean: make this an error *specifically* in teh Roslyn codebase.  There is 0 reason for us to ever have this sort of code ourselves :)

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi  In which case a Analyser and codefix is that is needed, and not a change to how the langauge works. :)

**Comment by CyrusNajmabadi** [comment]
I'm not asking for the langauge to change.  I'm asking for the feature that Neal has already written to simply run against the Roslyn codebase.   To be specific: Neal has a feature (behind some flag) that catches cases where people write code that is legal, but still totally broken.  We though that feature was running on the Roslyn codebase.  It isn't, and I and others have made this mistake in several places (see https://github.com/dotnet/roslyn/pull/14766).  This bug is about figuring out why Neal's check isn't working on the Roslyn codebase.

**Comment by gafter** [comment]
The reason you don't get errors for null comparisons involving ImmutableArray is because of this:
`public static bool operator ==(ImmutableArray<T>? left, ImmutableArray<T>? right); `
and its corresponding `!=`.
So if you have an `ImmutableArray<Foo> x`
The `(x != null)` uses this user-defined operator.
And we don't warn in that case, because we didn't lift anything, we simply used an operator that you declared.
It is only when we use a *lifted* operator for a `(s != null)` comparison that we warn.
